/*
 * Copyright (c) 2016 Qualcomm Technologies, Inc.
 * All Rights Reserved.
 * Confidential and Proprietary - Qualcomm Technologies, Inc.
 */
message MsgTag {
    required string        token       = 1;
    required MsgType       type        = 2;
    required MsgId         id          = 3;
    optional Error         error       = 4;
    optional bytes         payload     = 5;
}

enum Error {
    ERROR_SUCCESS          = 0;
    ERROR_GENERIC          = 1;
}
enum MsgType {
    MSG_TYPE_UNKNOWN       = 0;
    MSG_TYPE_REQUEST       = 1;
    MSG_TYPE_RESPONSE      = 2;
    MSG_TYPE_INDICATION    = 3;
}

enum MsgId {
    MSG_UNKNOWN                   = 0;
    MSG_RUN_COMMAND               = 1;
    MSG_GET_COMMANDS              = 2;
    MSG_PROPERTY_UPDATE_STARTED   = 3;
    MSG_PROPERTY_UPDATE_COMPLETED = 4;
}

/**
 * type: MSG_TYPE_REQUEST / MSG_TYPE_RESPONSE
 *   id: MSG_RUN_COMMAND
 */
message RunCommand
{
    required string        cmd_name    = 1;
    repeated string param              = 2;
}

/**
 * type: MSG_TYPE_REQUEST
 *   id: MSG_GET_COMMANDS
 */
message GetCommandsReq
{
}

/**
 * type: MSG_TYPE_RESPONSE
 *   id: MSG_GET_COMMANDS
 */
message GetCommandsRsp
{
    repeated Command       cmd           = 1;
}

message Command
{
    required string        cmd_name      = 1;
    required string        description   = 2;
}

/**
 * type: MSG_TYPE_INDICATION
 *   id: MSG_PROPERTY_UPDATE_STARTED
 */
message PropertyUpdateStarted
{
}

/**
 * type: MSG_TYPE_INDICATION
 *   id: MSG_PROPERTY_UPDATE_COMPLETED
 */
message PropertyUpdateCompleted
{
    repeated string        property_name    = 2;
}
